samples/extensions/dynamic_rendering/dynamic_rendering.cpp

```diff
@@ -0,0 +1,407 @@
+	specialization_info                  = vkb::initializers::specialization_info(1, specialization_map_entries.data(), sizeof(shadertype), &shadertype);
+	shader_stages[0].pSpecializationInfo = &specialization_info;
+	shader_stages[1].pSpecializationInfo = &specialization_info;
+
+	if (!enable_dynamic)
+	{
+		graphics_create.pNext      = VK_NULL_HANDLE;
+		graphics_create.renderPass = render_pass;
+	}
+
+	vkCreateGraphicsPipelines(get_device().get_handle(), VK_NULL_HANDLE, 1, &graphics_create, VK_NULL_HANDLE, &skybox_pipeline);
+
+	// Object rendering pipeline
+	shadertype = 1;
+
+	// Enable depth test and write
+	depth_stencil_state.depthWriteEnable = VK_TRUE;
+	depth_stencil_state.depthTestEnable  = VK_TRUE;
+	// Flip cull mode
+	rasterization_state.cullMode = VK_CULL_MODE_FRONT_BIT;
+	vkCreateGraphicsPipelines(get_device().get_handle(), VK_NULL_HANDLE, 1, &graphics_create, VK_NULL_HANDLE, &model_pipeline);
+}
+
+void DynamicRendering::create_render_pass_non_dynamic()
+{
+}
+
+void DynamicRendering::draw()
+{
+	ApiVulkanSample::prepare_frame();
+	submit_info.commandBufferCount = 1;
+	submit_info.pCommandBuffers    = &draw_cmd_buffers[current_buffer];
+	VK_CHECK(vkQueueSubmit(queue, 1, &submit_info, VK_NULL_HANDLE));
+	ApiVulkanSample::submit_frame();
+}
+
+void DynamicRendering::build_command_buffers()
+{
+	std::array<VkClearValue, 2> clear_values{};
+	clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+	clear_values[1].depthStencil = {0.0f, 0};
+
+	int i = -1;
+	for (auto &draw_cmd_buffer : draw_cmd_buffers)
+	{
+		i++;
+		auto command_begin = vkb::initializers::command_buffer_begin_info();
+		VK_CHECK(vkBeginCommandBuffer(draw_cmd_buffer, &command_begin));
+
+		auto draw_scene = [&] {
+			VkViewport viewport = vkb::initializers::viewport((float) width, (float) height, 0.0f, 1.0f);
+			vkCmdSetViewport(draw_cmd_buffer, 0, 1, &viewport);
+
+			VkRect2D scissor = vkb::initializers::rect2D(static_cast<int>(width), static_cast<int>(height), 0, 0);
+			vkCmdSetScissor(draw_cmd_buffer, 0, 1, &scissor);
+
+			// One descriptor set is used, and the draw type is toggled by a specialization constant
+			vkCmdBindDescriptorSets(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layout, 0, 1, &descriptor_set, 0, nullptr);
+
+			// skybox
+			vkCmdBindPipeline(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, skybox_pipeline);
+			draw_model(skybox, draw_cmd_buffer);
+
+			// object
+			vkCmdBindPipeline(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, model_pipeline);
+			draw_model(object, draw_cmd_buffer);
+
+			// UI
+			draw_ui(draw_cmd_buffer);
+		};
+
+		if (enable_dynamic)
+		{
+			VkRenderingAttachmentInfoKHR color_attachment_info = vkb::initializers::rendering_attachment_info();
+			color_attachment_info.imageView                    = swapchain_buffers[i].view;        // color_attachment.image_view;
+			color_attachment_info.imageLayout                  = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
```

Comment:
If my memory is right, and this isn't done by the framework, need pipeline barrier going in to transition PRESENT_SRC -> COLOR_ATTACHMENT_OPTIMAL here, prior to vkCmdBeginRenderingKHR, plus the opposite after vkCmdEndRenderingKHR in preparation of QueuePresent?

Reply:
DS image transitions were likely also handled with subpass-dependencies before, so this must also be done explicitly now.

Reply:
That's my understanding too. Afaik you need explicit barriers with dynamic rendering.

Reply:
Ohh man thanks Sascha and Christian!  I completely missed this.  We will address it in the morning.

Reply:
Thanks guys!  comments should be addressed now.